File: levelledmobs-plugin-bukkit/src/main/java/io/github/arcaneplugins/levelledmobs/bukkit/logic/function/process/action/impl/BroadcastMessageToServerAction.java
package io.github.arcaneplugins.levelledmobs.bukkit.logic.function.process.action.impl;

import io.github.arcaneplugins.levelledmobs.bukkit.config.translations.Message;
import io.github.arcaneplugins.levelledmobs.bukkit.logic.LogicHandler;
import io.github.arcaneplugins.levelledmobs.bukkit.logic.context.Context;
import io.github.arcaneplugins.levelledmobs.bukkit.logic.function.process.Process;
import io.github.arcaneplugins.levelledmobs.bukkit.logic.function.process.action.Action;
import java.util.Objects;
import net.kyori.adventure.text.Component;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.NotNull;
import org.spongepowered.configurate.CommentedConfigurationNode;
import org.spongepowered.configurate.ConfigurateException;

public class BroadcastMessageToServerAction extends Action {

    private final String requiredPermission;
    private String[] message = null;

    public BroadcastMessageToServerAction(
        final @NotNull Process process,
        final @NotNull CommentedConfigurationNode node
    ) {
        super(process, node);

        this.requiredPermission = getActionNode()
            .node("required-permission")
            .getString("");

        try {
            this.message = Objects.requireNonNull(
                getActionNode().node("message").getList(String.class),
                "message"
            ).toArray(new String[0]);
        } catch(ConfigurateException | NullPointerException ex) {
            throw new RuntimeException(ex);
        }
    }

    @Override
    public void run(Context context) {
        final String[] lines = new String[getMessage().length];
        for(int i = 0; i < lines.length; i++) {
            lines[i] = LogicHandler.replacePapiAndContextPlaceholders(getMessage()[i], context);
        }
        final Component msg = Message.formatMd(lines);

        for(final Player player : Bukkit.getOnlinePlayers()) {
            if(hasRequiredPermission() && !player.hasPermission(getRequiredPermission())) continue;
            player.sendMessage(msg);
        }
    }

    public boolean hasRequiredPermission() { return !requiredPermission.isEmpty(); }

    @NotNull
    public String getRequiredPermission() { return requiredPermission; }

    @NotNull
    public String[] getMessage() { return message; }
}

File: levelledmobs-plugin-bukkit/src/main/kotlin/io/github/arcaneplugins/levelledmobs/bukkit/logic/function/process/action/impl/BroadcastMessageToServerAction.kt
package io.github.arcaneplugins.levelledmobs.bukkit.logic.function.process.action.impl

import io.github.arcaneplugins.levelledmobs.bukkit.config.translations.Message
import io.github.arcaneplugins.levelledmobs.bukkit.debug.DebugCategory
import io.github.arcaneplugins.levelledmobs.bukkit.logic.LogicHandler.replacePapiAndContextPlaceholders
import io.github.arcaneplugins.levelledmobs.bukkit.logic.context.Context
import io.github.arcaneplugins.levelledmobs.bukkit.logic.function.process.Process
import io.github.arcaneplugins.levelledmobs.bukkit.logic.function.process.action.Action
import io.github.arcaneplugins.levelledmobs.bukkit.util.Log
import io.github.arcaneplugins.levelledmobs.bukkit.util.Log.debug
import org.bukkit.Bukkit
import org.spongepowered.configurate.CommentedConfigurationNode
import org.spongepowered.configurate.ConfigurateException

class BroadcastMessageToServerAction(
    val process: Process,
    val node: CommentedConfigurationNode
): Action(process, node) {
    private val requiredPermission: String = node.node("required-permission").getString("")
    val message: MutableList<String>

    init {
        try {
            this.message = node.node("message").getList(String::class.java)!!
        } catch (ex: ConfigurateException) {
            throw RuntimeException(ex)
        } catch (ex: NullPointerException) {
            throw RuntimeException(ex)
        }
    }

    override fun run(context: Context) {
        debug(DebugCategory.FUNCTIONS_GENERIC) {
            "BroadcastMessageToServerAction is running with message='$message'," +
                    "requiredPermission='$requiredPermission'."
        }

        val lines = mutableListOf<String>()
        for (line in message) {
            debug(DebugCategory.FUNCTIONS_GENERIC) {
                "BroadcastMessageToServerAction BEFORE replacePapiAndContextPlaceholders LINE: $line"
            }
            lines.add(replacePapiAndContextPlaceholders(line, context))
            debug(DebugCategory.FUNCTIONS_GENERIC) {
                "BroadcastMessageToServerAction line added to lines, new lines size: ${lines.size}"
            }
        }
        debug(DebugCategory.FUNCTIONS_GENERIC) {
            "BroadcastMessageToServerAction replacePapiAndContextPlaceholders lines: $lines"
        }

        val msg = Message.formatMd(lines)

        for (player in Bukkit.getOnlinePlayers()) {
            debug(DebugCategory.FUNCTIONS_GENERIC) {"BroadcastMessageToServerAction is checking perms of ${player.name}"}
            if (requiresPermission() && !player.hasPermission(requiredPermission)) continue
            debug(DebugCategory.FUNCTIONS_GENERIC) {"BroadcastMessageToServerAction is sending msg to ${player.name}"}
            player.sendMessage(msg)
        }
    }

    private fun requiresPermission(): Boolean {
        return requiredPermission.isNotEmpty()
    }
}